### PR TITLE
libs/wolfssl: add support for all configure options

### PR DESCRIPTION
**Pull request by commodo**
WolfSSL has quite a few configuration options, which is good for managing
size and desired options.
Up until now, only a few have been specifically selected by the OpenWrt
build system.

This change adds support for almost all of them.
All is not possible since some are arch dependent.

Admittedly, this breaks some existing stuff in OpenWrt, because some options
have been renamed for convenience, and some defaults may have changed.
The defaults are the same as in WolfSSL.

But this method is simple to maintain moving forward, because all that is
required is to grep all `enable-` options from wolfSSL's
configure script and update the WOLFSSL_DEFAULT_{ON/OFF}_OPTS variables.

A few of the default values enabled/disabled are the same as in WolfSSL.
Others have been adapted for the OpenWrt case: example ipv6 has been tied
to CONFIG_IPV6.

This change also supports grouping config symbols.
If a group of configure symbols should be enabled/disbabled under 1 symbol,
they can be specified via `sym1,sym2,...,symN`.
Only sym1 will appear in menuconfig and enable/disable all N symbols.
Note that `symX` symbols should all be valid `--enable-symX` symbols for
wolfSSL.

Dependencies are supported as well, and can be neatly grouped in
menuconfig.
But order is important to display them properly.
To do this, a [group] symbol needs to be added to the
WOLFSSL_GROUP_DEP_SYMBOLS variable.
Then it needs to be defined in one of the  WOLFSSL_DEFAULT_{ON/OFF}_OPTS
variables.
Example:
        WOLFSSL_GROUP_DEP_SYMBOLS:= sym1
        WOLFSSL_DEFAULT_ON_OPTS:= sym1 sym1:sym2
Note that sym1 needs to be defined as well, and has to be defined before
`sym1:sym2`. All symX symbols need to be valid `--enable-symX` symbols for
wolfSSL.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

### CODE REVIEW

**Comment by commodo** [comment]
@dangowrt  ping
took me a while to calibrate these; seems to work on my VM;
there's still room for plenty of fine-tuning with defaults and stuff, but this should be a good base to start from

**Comment by commodo** [comment]
changelog v1 -> v2:
* bumped PKG_RELEASE

**Comment by dangowrt** [comment]
nice one!
I guess we should auto-detect the assembly options for each platform,
we do know all target ABI properties at compile time.
See eg. `multimedia/ffmpeg/Makefile`.
However, I'm also ok with merging this as-is for now and I'd change
that in a follow-up step.

**Comment by commodo** [comment]
please test this first on your setup;

atm, auto-detecting assembly opts seems tricky;
because for ARM you need to go by CPU type/sub-type, not by ARCH;
some CPUs have these instructions; some don't;
i didn't find (yet) to which target optimization option these SHA instructions belong to;
[ I didn't search too in-depth yet ]
we'd also need to check which GCCs have these instructions too;
when I enabled them for ARM I got linker errors for all these assembly symbols;

**Comment by dangowrt** [comment]
Knowing ABI details is possible just like we do for ffmpeg and others,
eg. to detect whether to build floating-point and vector stuff for VFPx
or NEON. From what I can tell after a short look at wolfSSL, the only
thing we need to know is whether we are on x86_64 and should hence
support all the intelasm, intelrand and aesni stuff which **may** be
supported by the CPU. Auto-detecting CPU features at runtime should
work on x86-platforms as one would hardly ever have CPU-type specific
binary builds on that platform. Other than that, all I can find is
support for ARMv8 which we should enable on arm64 builds as all needed
for that seems 64-bit ARM ABI.

File: package/libs/wolfssl/Makefile [hunk 1]
```diff
@@ -23,14 +23,83 @@ PKG_BUILD_PARALLEL:=1
...
+	tls13:tls13-draft23 \
+	distro 32bit singlethreaded dtls postauth hrrcookie sctp mcast openssh \
+	nginx haproxy bump sniffer signal maxstrength leanpsk leantls \
+	lowresource bigcache hugecache smallcache savesession savecert writedup \
+	atomicuser pkcallbacks xilinx camellia md2 nullcipher ripemd blake2 \
+	sessioncerts keygen certgen certreq certext sep hkdf x963kdf dsa \
+	compkey psk tlsv10 sslv3 trackmemory des3 idea arc4 cmac xts \
+	webserver fips hc128 rabbit sha224 sha3 qsh sni maxfragment alpn \
+	truncatedhmac renegotiation-indication secure-renegotiation \
+	session-ticket tlsx earlydata pkcs7 ssh scep srp indef smallstack \
```
Comment (cotequeiroz) [inline package/libs/wolfssl/Makefile]: cannot enable earlydata without enabling tls13 (configure.ac: 2646)
cannot enable earlydata without enabling session tickets and/or PSK (configure.ac: 2650)
pkcs7 needs: rsa, sha, and either aes or 3des (configure.ac: 3905-3917). 3des is always on (stunnel)
scep also needs either aes or 3des.
ssh depends on aesgcm, ecc, sha, and sha512 (configure.ac: 4049-4052).

**Review by cotequeiroz — COMMENTED** [review]
There are many option implications that need to be picked up from code in `configure.ac`.  I have pointed out the non-fatal ones that I dealt with in #1069 (trying to turn off ECC is what got me started), and listed here the ones that generate build errors.  I mention the configure.ac line numbers for the ones that generate errors.
I think most of them can be dealt with your clever dependency system.
Cheers

File: package/libs/wolfssl/Makefile [hunk 2]
```diff
@@ -50,109 +119,68 @@ wolfSSL (formerly CyaSSL) is an SSL library optimized for small
...
 
 # --enable-stunnel needed for OpenSSL API compatibility bits
+# --disable-opensslcoexist - requires that OpenSSL exist to build
+# --disable-optflags - we want to use OpenWrt's CFLAGS
+# --disable-stacksize - stack size info on examples (we disable examples)
+# --disable-selftest - requires CAVP selftest license
 CONFIGURE_ARGS += \
 	--enable-opensslextra \
-	--enable-sni \
 	--enable-stunnel \
```
[the remark this answers is not in the document]
Comment (cotequeiroz) [reply package/libs/wolfssl/Makefile]: sni is a requirement for ustream-ssl, so if we make this optional here, we need to add a depend flag there.

**Comment by commodo** [comment]
@cotequeiroz 
thanks for the thorough review ; i'll take a look these days;
i'm a bit swamped with other work; but i'll try to get there asap

**Comment by blogic** [comment]
please reopen when you had time to fix up the comments

**Pull request closed without merging**